### DISCUSSION TITLE
Upcoming purge feature

### DISCUSSION THREAD

**Discussion by mimcdevitt** [apache/cloudstack#9365] [Q&A] [answered]
Hello,  For the upcoming [Purge Expunged Resources](https://cwiki.apache.org/confluence/display/CLOUDSTACK/Purge+Expunged+Resources) feature.  Is there a list of database tables that will be impacted as part of this feature.  We have had some performance issues due to large number of entries in some of the cloudstack database tables and hoping the purge feature will be the permanent solution.

**Comment by shwstppr** [comment, marked as answer]
Hi @mimcdevitt, currently the feature supports purging of only the VirtualMachines and their associated resources. Here is a list of DB tables that will be affected:

- vm_instance
- user_vm
- user_vm_details
- nics
- vm_snapshots
- console_proxy
- domain_router
- autoscale_vmgroup_vm_map
- cmd_exec_log
- elastic_lb_vm_map
- load_balancer_vm_map
- nic_secondary_ips
- op_ha_work
- op_it_work
- op_router_monitoring_services
- port_forwarding_rules
- secondary_storage_vm
- user_ip_address
- vm_work_job
- volumes
- volume_details
- volume_host_ref
- volume_store_ref
- snapshot_schedule
- snapshots
- snapshot_store_ref
- snapshot_schedule
- nics
- nic_details
- nic_extra_dhcp_options
- inline_load_balancer_nic_map
- console_session

I hope I didn't miss any. You may refer this class to check, https://github.com/apache/cloudstack/blob/main/server/src/main/java/org/apache/cloudstack/resource/ResourceCleanupServiceImpl.java

Keep in mind the above list contains tables such as volume, snapshots, etc but entries from them will be purged only when the volume is linked to a VM entry that is getting purged. Since DATA volumes when deleted are delinked from the VM their entries do not get purged.

**Reply by mimcdevitt** [reply]
@shwstppr Thank you very much for your insights and information.
